File: skills/salomonhenao/veritier-fact-checking/javascript/webhooks/webhook_receiver.mjs
/**
 * Webhook Receiver with HMAC-SHA256 Verification - Veritier Example (JavaScript)
 * ================================================================================
 * A minimal Express server that receives async webhook deliveries from Veritier
 * and verifies their HMAC-SHA256 signatures before processing.
 *
 * When you enable webhooks in your Veritier dashboard and set `use_webhook: true`
 * in API requests, results are delivered asynchronously to this endpoint.
 *
 * Setup:
 *   1. npm install express dotenv
 *   2. Set VERITIER_WEBHOOK_SECRET in your .env (the vtsec_... value from the dashboard)
 *   3. node webhook_receiver.mjs
 *   4. The server listens on http://localhost:5050/webhooks/veritier
 *
 * For production:
 *   - Use HTTPS (required by Veritier for non-localhost URLs)
 *   - Deploy behind a reverse proxy (nginx, Caddy, etc.)
 *
 * Full webhook docs: https://veritier.ai/docs
 */

import "dotenv/config";
import express from "express";
import { createHmac, timingSafeEqual } from "node:crypto";

const app = express();
const PORT = 5050;

const WEBHOOK_SECRET = process.env.VERITIER_WEBHOOK_SECRET || "";

if (!WEBHOOK_SECRET) {
  console.warn("⚠ Warning: VERITIER_WEBHOOK_SECRET is not set.");
  console.warn("  Configure a webhook at https://veritier.ai/dashboard to get your secret.");
}

// IMPORTANT: Use raw body for HMAC verification, NOT parsed JSON
// Express must capture the raw bytes before any JSON middleware
app.use(
  "/webhooks/veritier",
  express.raw({ type: "application/json" })
);
app.use(express.json());

app.post("/webhooks/veritier", (req, res) => {
  /**
   * Receive and verify a Veritier webhook delivery.
   *
   * Security flow:
   *   1. Read the raw request body BEFORE any JSON parsing
   *   2. Compute HMAC-SHA256 of the raw bytes using your webhook secret
   *   3. Compare against the X-Veritier-Signature header (timing-safe)
   *   4. Only then parse and process the payload
   */
  const signature = req.headers["x-veritier-signature"] || "";

  if (!WEBHOOK_SECRET) {
    console.error("✗ Webhook secret not configured - rejecting request");
    return res.status(500).json({ error: "Webhook secret not configured" });
  }

  // ── Step 1: Verify the signature ────────────────────────────────────
  // req.body is a Buffer here because of express.raw() middleware
  const rawBody = req.body;

  const expectedSignature =
    "vtsec_" +
    createHmac("sha256", WEBHOOK_SECRET)
      .update(rawBody)
      .digest("hex");

  // Timing-safe comparison prevents signature oracle attacks
  try {
    const sigBuffer = Buffer.from(signature, "utf8");
    const expectedBuffer = Buffer.from(expectedSignature, "utf8");

    if (
      sigBuffer.length !== expectedBuffer.length ||
      !timingSafeEqual(sigBuffer, expectedBuffer)
    ) {
      console.error("✗ Invalid webhook signature - rejecting request");
      return res.status(401).json({ error: "Invalid signature" });
    }
  } catch {
    console.error("✗ Signature comparison failed - rejecting request");
    return res.status(401).json({ error: "Invalid signature" });
  }

  // ── Step 2: Parse and process the payload ───────────────────────────
  const payload = JSON.parse(rawBody.toString("utf8"));
  const transactionId = payload.transaction_id || "unknown";
  const results = payload.results || [];

  const icons = { true: "✅", false: "❌", null: "❓" };

  console.log(`\n${"─".repeat(50)}`);
  console.log(`✓ Webhook received - Transaction: ${transactionId}`);
  console.log(`  Claims verified: ${results.length}`);

  for (const r of results) {
    const icon = icons[String(r.verdict)] || "❓";
    console.log(`  ${icon} ${r.claim} → ${r.verdict}`);
  }

  console.log(`${"─".repeat(50)}\n`);

  return res.json({ status: "ok" });
});

app.get("/health", (_req, res) => {
  res.json({ status: "ok" });
});

app.listen(PORT, () => {
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
  console.log("  Veritier Webhook Receiver");
  console.log(`  Listening on http://localhost:${PORT}/webhooks/veritier`);
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");
});

File: skills/salomonhenao/veritier-fact-checking/python/webhooks/webhook_receiver.py
#!/usr/bin/env python3
"""
Webhook Receiver with HMAC-SHA256 Verification - Veritier Example (Python)
===========================================================================
A minimal Flask server that receives async webhook deliveries from Veritier
and verifies their HMAC-SHA256 signatures before processing.

When you enable webhooks in your Veritier dashboard and set `use_webhook: true`
in API requests, results are delivered asynchronously to this endpoint.

Setup:
  1. pip install flask python-dotenv
  2. Set VERITIER_WEBHOOK_SECRET in your .env (the vtsec_... value from the dashboard)
  3. python webhook_receiver.py
  4. The server listens on http://localhost:5050/webhooks/veritier

For production:
  - Use HTTPS (required by Veritier for non-localhost URLs)
  - Deploy behind a reverse proxy (nginx, Caddy, etc.)
  - Use a WSGI server (gunicorn, waitress)

Full webhook docs: https://veritier.ai/docs
"""

import hmac
import hashlib
import os
from dotenv import load_dotenv
from flask import Flask, request, abort, jsonify

load_dotenv()

app = Flask(__name__)

WEBHOOK_SECRET = os.getenv("VERITIER_WEBHOOK_SECRET", "")

if not WEBHOOK_SECRET:
    print("⚠ Warning: VERITIER_WEBHOOK_SECRET is not set.")
    print("  Configure a webhook at https://veritier.ai/dashboard to get your secret.")


@app.route("/webhooks/veritier", methods=["POST"])
def veritier_webhook():
    """
    Receive and verify a Veritier webhook delivery.

    Security flow:
      1. Read the raw request body BEFORE any JSON parsing
      2. Compute HMAC-SHA256 of the raw bytes using your webhook secret
      3. Compare against the X-Veritier-Signature header (timing-safe)
      4. Only then parse and process the payload
    """
    signature = request.headers.get("X-Veritier-Signature", "")

    if not WEBHOOK_SECRET:
        print("✗ Webhook secret not configured - rejecting request")
        abort(500)

    # ── Step 1: Verify the signature ────────────────────────────────────
    # IMPORTANT: Use request.data (raw bytes), NOT request.get_json()
    # Re-serializing JSON may change whitespace/key order and break the HMAC
    raw_body = request.data

    expected_signature = "vtsec_" + hmac.new(
        key=WEBHOOK_SECRET.encode("utf-8"),
        msg=raw_body,
        digestmod=hashlib.sha256,
    ).hexdigest()

    # Timing-safe comparison prevents signature oracle attacks
    if not hmac.compare_digest(signature, expected_signature):
        print("✗ Invalid webhook signature - rejecting request")
        abort(401)

    # ── Step 2: Parse and process the payload ───────────────────────────
    payload = request.get_json()

    transaction_id = payload.get("transaction_id", "unknown")
    results = payload.get("results", [])

    print(f"\n{'─' * 50}")
    print(f"✓ Webhook received - Transaction: {transaction_id}")
    print(f"  Claims verified: {len(results)}")

    for res in results:
        verdict = res.get("verdict")
        icon = {True: "✅", False: "❌", None: "❓"}.get(verdict, "❓")
        print(f"  {icon} {res.get('claim')} → {verdict}")

    print(f"{'─' * 50}\n")

    return jsonify({"status": "ok"}), 200


@app.route("/health", methods=["GET"])
def health():
    return jsonify({"status": "ok"}), 200


if __name__ == "__main__":
    print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print("  Veritier Webhook Receiver")
    print("  Listening on http://localhost:5050/webhooks/veritier")
    print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n")
    app.run(host="0.0.0.0", port=5050, debug=True)
